proposals/NNNN-lifetime-dependency.md

```diff
@@ -0,0 +1,481 @@
+* `ResultType` must be `~Escapable`.
+
+If the `ArgType` is `Escapable`, the return value will have a new scoped dependency on the argument.
+(This is the only possibility, as an `Escapable` value cannot have an existing lifetime dependency,
+so we cannot copy it.)
+A scoped dependency ensures the argument will not be destroyed while the result is alive.
+Also, access to the argument will be restricted for the lifetime of the result following Swift's usual exclusivity rules:
+
+* A `borrowing` parameter-convention extends borrowing access, prohibiting mutations of the argument.
+* An `inout` parameter-convention extends mutating access, prohibiting any access to the argument.
+* A `consuming` parameter-convention is illegal, since that ends the lifetime of the argument immediately.
+
+If the `ArgType` is `~Escapable`, then it can have a pre-existing lifetime dependency.
+In this case, the semantics of `@dependsOn()` are slightly different:
+* A `consuming` parameter-convention will copy the lifetime dependency from the argument to the result
+* A `borrowing` or `inout` parameter-convention can either copy the lifetime dependency or create a new scoped lifetime dependency.
+  In this case, for reasons explained earlier, we default to copying the lifetime dependency.
+  If a scoped lifetime dependency is needed, it can be explicitly requested by adding the `scoped` keyword:
+  
+```swift
+func f(arg: borrowing ArgType) -> @dependsOn(scoped arg) ResultType
+```
+
+**Methods:** Similar rules apply to `self` lifetime dependencies on methods.
+Given a method of this form:
+
+```swift
+<mutation-modifier> func method(... args ...) -> @dependsOn(self) ResultType
+```
+
+The behavior depends as above on the mutation-modifier and whether the defining type is `Escapable` or `~Escapable`.
+
+**Initializers:** An initializer can define a lifetime dependency on one of its arguments.
+In this case, we use the same rules same as for “Functions” above
+by using the convention that initializers can be viewed as functions that return `Self`:
+
+```swift
+init(arg: <parameter-convention> ArgType) -> @dependsOn(arg) Self
+```
+
+### Implicit Lifetime Dependencies
+
+The syntax above allows developers to explicitly annotate lifetime dependencies in their code.
+But because the possibilities are limited, we can usually allow the compiler to infer a suitable dependency.
+The detailed rules are below, but generally we require that the return type be `~Escapable` and that there be one “obvious” source for the dependency.
+
+In particular, we can infer a lifetime dependency on `self` for any method that returns a `~Escapable` type.
+As above, the details vary depending on whether `self` is `Escapable` or `~Escapable`:
+
+```swift
+struct NonescapableType: ~Escapable { ... }
+struct EscStruct {
+  func f1(...) -> /* @dependsOn(self) */ NonescapableType
```

Comment:
We don't infer when mutation modifier is absent

Reply:
Is there a technical concern?  Or is this just how the current implementation works?

I discussed this with @atrick and I thought we concluded it would be reasonable to accept this without an explicit mutation modifier.

Reply:
As per latest on this, I'll update the implementation